src/lobject.h

```diff
@@ -54,19 +54,43 @@ typedef uint16_t LuaType;
 ** we care about from a performance point of view - if any
 ** other types appear then they are all treated as ANY
 **/
+
 typedef enum {
-  RAVI_TANY = 0,      /* Lua dynamic type */
-  RAVI_TNUMINT = 1,    /* integer number */
-  RAVI_TNUMFLT,        /* floating point number */
-  RAVI_TARRAYINT,      /* array of ints */
-  RAVI_TARRAYFLT,      /* array of doubles */
-  RAVI_TFUNCTION,      /* Lua or C Function */
-  RAVI_TTABLE,         /* Lua table */
-  RAVI_TSTRING,        /* string */
-  RAVI_TNIL,           /* NIL */
-  RAVI_TBOOLEAN,       /* boolean */
-  RAVI_TUSERDATA       /* userdata or lightuserdata */
-} ravitype_t;
+  RAVI_TI_NIL,
+  RAVI_TI_FALSE,
+  RAVI_TI_TRUE,
+  RAVI_TI_INTEGER,
+  RAVI_TI_FLOAT,
+  RAVI_TI_INTEGER_ARRAY,
+  RAVI_TI_FLOAT_ARRAY,
+  RAVI_TI_TABLE,
+  RAVI_TI_STRING,
+  RAVI_TI_FUNCTION,
+  RAVI_TI_USERDATA,
+  RAVI_TI_OTHER
+} ravi_type_index;
+
+typedef uint32_t ravi_type_map;
+
+#define RAVI_TM_NIL (((ravi_type_map)1)<<RAVI_TI_NIL)
+#define RAVI_TM_FALSE (((ravi_type_map)1)<<RAVI_TI_FALSE)
+#define RAVI_TM_TRUE (((ravi_type_map)1)<<RAVI_TI_TRUE)
+#define RAVI_TM_INTEGER (((ravi_type_map)1)<<RAVI_TI_INTEGER)
+#define RAVI_TM_FLOAT (((ravi_type_map)1)<<RAVI_TI_FLOAT)
+#define RAVI_TM_INTEGER_ARRAY (((ravi_type_map)1)<<RAVI_TI_INTEGER_ARRAY)
+#define RAVI_TM_FLOAT_ARRAY (((ravi_type_map)1)<<RAVI_TI_FLOAT_ARRAY)
+#define RAVI_TM_TABLE (((ravi_type_map)1)<<RAVI_TI_TABLE)
+#define RAVI_TM_STRING (((ravi_type_map)1)<<RAVI_TI_STRING)
+#define RAVI_TM_FUNCTION (((ravi_type_map)1)<<RAVI_TI_FUNCTION)
+#define RAVI_TM_USERDATA (((ravi_type_map)1)<<RAVI_TI_USERDATA)
+#define RAVI_TM_OTHER (((ravi_type_map)1)<<RAVI_TI_OTHER)
+
+#define RAVI_TM_FALSISH (RAVI_TM_NIL | RAVI_TM_FALSE)
+#define RAVI_TM_TRUISH (~RAVI_TM_FALSISH)
+#define RAVI_TM_BOOLEAN (RAVI_TM_FALSE | RAVI_TM_TRUE)
+#define RAVI_TM_NUMBER (RAVI_TM_INTEGER | RAVI_TM_FLOAT)
+#define RAVI_TM_INDEXABLE (RAVI_TM_INTEGER_ARRAY | RAVI_TM_FLOAT_ARRAY | RAVI_TM_TABLE)
+#define RAVI_TM_ANY (~0)
```

Comment:
Thinking aloud here. Instead of the index enum and these macros - could we not extend the original ravitype_t enum with the additional union types?

Reply:
With this structure the old enum values would still require to get different values. The same could be achieved with preprocessor defines. Like
```c
typedef ravi_type_map ravitype_t; // Note the suffix _t is reserved by posix.
#define RAVI_TANY RAVI_TM_ANY	  /* Lua dynamic type */
#define RAVI_TNUMINT RAVI_TM_INTEGER /* integer number */
#define RAVI_TNUMFLT RAVI_TM_FLOAT	  /* floating point number */
#define RAVI_TARRAYINT RAVI_TM_INTEGER_ARRAY	  /* array of ints */
#define RAVI_TARRAYFLT RAVI_TM_FLOAT_ARRAY	  /* array of doubles */
#define RAVI_TFUNCTION (RAVI_TM_FUNCTION | RAVI_TM_NIL)	  /* Lua or C Function */
#define RAVI_TTABLE RAVI_TM_TABLE	  /* Lua table */
#define RAVI_TSTRING (RAVI_TM_FUNCTION | RAVI_TM_NIL)	  /* string */
#define RAVI_TNIL RAVI_TM_NIL	  /* NIL */
#define RAVI_TBOOLEAN (RAVI_TM_BOOLEAN | RAVI_TM_NIL)	  /* boolean */
#define RAVI_TUSERDATA (RAVI_TM_USERDATA | RAVI_TM_NIL)	  /* userdata or lightuserdata */
```
Not that the size of `ravitype_t` changed and the values of the old enum value. Since these values don't create a real enum I would not use an enum here.

Reply:
We could perhaps also squeeze the into a byte - the way Lua types are.

Reply:
> Since these values don't create a real enum I would not use an enum here.

What do you mean by 'real enum' ?

Reply:
No, this can not be queezed down to one bytes since this is a bitmap and needs more than 8 bits, since there are more than 8 types. You normally don't do with real enums:
```c
typedef enum {
   flag1 = 1,
   flag2 = 2,
   flag3 = 4,
} flags;
flags flag = flag1 | flag2 | flag3;
```
The value of `flag` is not any more in the enum.
For example `b and 1` needs to be of type `RAVI_TM_NIL | RAVI_TM_FALSE | RAVI_TM_INTEGER` since the left side can be `RAVI_TM_NIL | RAVI_TM_FALSE` if it is any other type the right side would be used. And the right side is `RAVI_TM_INTEGER` and we need to combine those.

Reply:
I think we can still use enums rather than macros ... I  prefer them because they show up meaningfully in the debugger, but macros make it harder to debug. But I am still trying to figure out how best to proceed and whether this is the best way.